x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/dataframe/stats/MemoryUsage.java

```diff
@@ -0,0 +1,115 @@
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.core.ml.utils.ToXContentParams;
+
+import java.io.IOException;
+import java.time.Instant;
+import java.util.Objects;
+
+public class MemoryUsage implements Writeable, ToXContentObject {
+
+    public static final String TYPE_VALUE = "analytics_memory_usage";
+
+    public static final ParseField TYPE = new ParseField("type");
+    public static final ParseField JOB_ID = new ParseField("job_id");
+    public static final ParseField TIMESTAMP = new ParseField("timestamp");
+    public static final ParseField PEAK_USAGE_BYTES = new ParseField("peak_usage_bytes");
+
+    public static final ConstructingObjectParser<MemoryUsage, Void> STRICT_PARSER = createParser(false);
+    public static final ConstructingObjectParser<MemoryUsage, Void> LENIENT_PARSER = createParser(true);
+
+    private static ConstructingObjectParser<MemoryUsage, Void> createParser(boolean ignoreUnknownFields) {
+        ConstructingObjectParser<MemoryUsage, Void> parser = new ConstructingObjectParser<>(TYPE_VALUE,
+            ignoreUnknownFields, a -> new MemoryUsage((String) a[0], (Instant) a[1], (long) a[2]));
+
+        parser.declareString((bucket, s) -> {}, TYPE);
+        parser.declareString(ConstructingObjectParser.constructorArg(), JOB_ID);
+        parser.declareField(ConstructingObjectParser.constructorArg(),
+            p -> TimeUtils.parseTimeFieldToInstant(p, TIMESTAMP.getPreferredName()),
+            TIMESTAMP,
+            ObjectParser.ValueType.VALUE);
+        parser.declareLong(ConstructingObjectParser.constructorArg(), PEAK_USAGE_BYTES);
+        return parser;
+    }
+
+    private final String jobId;
+    private final Instant timestamp;
+    private final long peakUsageBytes;
+
+    public MemoryUsage(String jobId, Instant timestamp, long peakUsageBytes) {
+        this.jobId = Objects.requireNonNull(jobId);
+        this.timestamp = Instant.ofEpochMilli(ExceptionsHelper.requireNonNull(timestamp, TIMESTAMP).toEpochMilli());
+        this.peakUsageBytes = peakUsageBytes;
+    }
+
+    public MemoryUsage(StreamInput in) throws IOException {
+        jobId = in.readString();
+        timestamp = in.readInstant();
+        peakUsageBytes = in.readVLong();
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        out.writeString(jobId);
+        out.writeInstant(timestamp);
+        out.writeVLong(peakUsageBytes);
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        if (params.paramAsBoolean(ToXContentParams.FOR_INTERNAL_STORAGE, false)) {
+            builder.field(TYPE.getPreferredName(), TYPE_VALUE);
+            builder.field(JOB_ID.getPreferredName(), jobId);
+        }
+        builder.timeField(TIMESTAMP.getPreferredName(), TIMESTAMP.getPreferredName() + "_string", timestamp.toEpochMilli());
```

Comment:
It would be good to add a comment that the reason for rounding to millisecond accuracy is that the XContent representation rounds to millisecond accuracy and it makes debugging hard if the internal accuracy is greater.